### PR TITLE
Migration guide for JAX-RS changes

### LINKED ISSUE #15454: Update migration guide with the changes that need to be done for developers using JAX-RS in their extensions
### Description

The move to Resteasy Reactive is introducing a few breaking changes that need to be considered when upgrading Keycloak:

* No longer possible to use inner classes (non-static nested classes) but change them to `static` nested classes
* No longer possible to inject contextual objects using `@Context` annotation but leverage our `Resteasy` SPI to obtain instances (still under investigation)
* Include a `META-INF/beans.xml` file in the provider JAR to make it possible to index/discover JAX-RS resources so that Quarkus can properly generate bytecode for proxies and still allow context injection. (still under investigation)

### Discussion

https://github.com/keycloak/keycloak/issues/10713

### Motivation

https://github.com/keycloak/keycloak/issues/10713

### Details

_No response_

### PR DESCRIPTION
**Pull request by pedroigor**
Related to https://github.com/keycloak/keycloak/issues/15605.

Closes https://github.com/keycloak/keycloak/issues/15454

### CODE REVIEW

File: upgrading/topics/keycloak/changes-21_0_0.adoc
Comment (andymunro) [inline upgrading/topics/keycloak/changes-21_0_0.adoc]:
```suggestion
You should review your custom JAX-RS (sub)resources in order to obtain any contextual data from a `org.keycloak.common.util.ResteasyProvider`
```

File: upgrading/topics/keycloak/changes-21_0_0.adoc
Comment (andymunro) [inline upgrading/topics/keycloak/changes-21_0_0.adoc]:
```suggestion
KeycloakSession session = // obtain the session, which is usually available when creating a custom provider from a factory
```

File: upgrading/topics/keycloak/changes-21_0_0.adoc
Comment (andymunro) [inline upgrading/topics/keycloak/changes-21_0_0.adoc]:
```suggestion
In case you have no access to a `KeycloakSession` instance when invoking a JAX-RS resource method, you can obtain
```

File: upgrading/topics/keycloak/changes-21_0_0.adoc
Comment (andymunro) [inline upgrading/topics/keycloak/changes-21_0_0.adoc]:
```suggestion
You can still resolve contextual data from the JAX-RS runtime by:
```

File: upgrading/topics/keycloak/changes-21_0_0.adoc
Comment (andymunro) [inline upgrading/topics/keycloak/changes-21_0_0.adoc]:
```suggestion
You need to add a `META-INF/beans.xml` to the JAR file where your custom providers are packaged. Otherwise, they are not recognized by the server
```

File: release_notes/topics/21_0_0.adoc
Comment (andymunro) [inline release_notes/topics/21_0_0.adoc]:
```suggestion
See the link:{upgradingguide_link}[{upgradingguide_name}] for more details about how to
```

**Review by andymunro — CHANGES_REQUESTED** [review]
@pedroigor A few suggestions for change.

**Comment by pedroigor** [comment]
@andymunro Thanks. I'm accepting your suggestions now and I'll be squashing the commits once the upstream change is merged.

**Comment by ahus1** [comment]
@pedroigor - we're currently trying to merge all open PRs before we move over the docs. 

This PR is still in draft an one of the issues mentioned above is now merged. So wonder if this PR can be merged within the next days, or if it should rather stay open and will be merged after the move of the docs. 

Thanks!

**Comment by ahus1** [comment]
The documentation has been moved to [the main Keycloak repository](https://github.com/keycloak/keycloak/tree/main/docs/documentation).

Please move these changes to the main repository by creating a new pull request there. 

Thanks!

**Pull request closed without merging**
